src/MPDevice.cpp

```diff
@@ -1201,51 +1460,351 @@ void MPDevice::tagPointedNodes()
+                {
+                    qWarning() << "tagPointedNodes: couldn't find data child node with address" << tempChildAddress.toHex() << "in our list";
+                    return_bool = false;
+
+                    /* TODO: fix prev child / parent node */
+                    if (tempChildAddress == tempParentNodePt->getStartChildAddress())
+                    {
+                        // first child
+                    }
+                    else
+                    {
+                        // fix prev child
+                    }
+
+                    /* Loop to next parent */
+                    tempChildAddress = MPNode::EmptyAddress;
+                }
+                else if (tempNextChildNodePt->getPointedToCheck() == true)
+                {
+                    /* Linked chain loop detected */
+                    if (tempChildAddress == tempParentNodePt->getStartChildAddress())
+                    {
+                        qCritical() << "tagPointedNodes: data child node already pointed to: parent node with address" << tempParentAddress.toHex() << " points to child node with address" << tempChildAddress.toHex();
+                    }
+                    else
+                    {
+                        qCritical() << "tagPointedNodes: data child node loop has been detected: child node with address" << tempChildNodePt->getAddress().toHex() << " points to child node with address" << tempChildAddress.toHex();
+                    }
+
+                    /* Stop looping */
+                    return false;
+                }
+                else
+                {
+                    /* Set correct pointed node */
+                    tempChildNodePt = tempNextChildNodePt;
+
+                    /* Tag child */
+                    tempChildNodePt->setPointedToCheck();
+
+                    /* Loop to next possible child */
+                    tempChildAddress = tempChildNodePt->getNextChildDataAddress();
+                }
+            }
+
+            /* get next parent address */
+            tempParentAddress = tempParentNodePt->getNextParentAddress();
+        }
+    }
+
+    return return_bool;
 }
 
-bool MPDevice::checkLoadedNodes()
+bool MPDevice::checkLoadedNodes(bool repairAllowed)
 {
+    QList<QByteArray>::iterator addresslist_iterator;
+    QByteArray temp_pnode_address, temp_cnode_address;
+    MPNode* temp_pnode_pointer;
+    MPNode* temp_cnode_pointer;
     QList<MPNode *>::iterator i;
+    bool return_bool;
 
     qInfo() << "Checking database...";
 
-    /* Tag pointed nodes */
-    tagPointedNodes();
+    /* Tag pointed nodes, also detects DB errors */
+    return_bool = tagPointedNodes(repairAllowed);
 
     /* Scan for orphan nodes */
     quint32 nbOrphanParents = 0;
     quint32 nbOrphanChildren = 0;
+    quint32 nbOrphanDataParents = 0;
+    quint32 nbOrphanDataChildren = 0;
     for (i = loginNodes.begin(); i != loginNodes.end(); i++)
```

Comment:
Although it works, it is ugly because you need to dereference with (*i)-> to use the object pointer.
Use instead c++11 for range loop:
```
for (auto &i: loginNodes)
{
   if (i->getPointedToCheck())
...
```